File: src/main/java/com/saomc/screens/window/ScreenGUI.java
package com.saomc.screens.window;

import com.saomc.GLCore;
import com.saomc.colorstates.CursorStatus;
import com.saomc.resources.StringNames;
import com.saomc.screens.Elements;
import com.saomc.screens.ParentElement;
import com.saomc.screens.buttons.Actions;
import com.saomc.util.ColorUtil;
import com.saomc.util.OptionCore;
import net.minecraft.client.gui.GuiScreen;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.lwjgl.BufferUtils;
import org.lwjgl.LWJGLException;
import org.lwjgl.input.Cursor;
import org.lwjgl.input.Mouse;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

@SideOnly(Side.CLIENT)
public abstract class ScreenGUI extends GuiScreen implements ParentElement {

    private static final float ROTATION_FACTOR = 0.25F;
    protected static CursorStatus CURSOR_STATUS = CursorStatus.SHOW;
    protected final List<Elements> elements;
    private final Cursor emptyCursor;
    private int mouseX, mouseY;
    private int mouseDown;
    private float mouseDownValue;
    private float[] rotationYaw, rotationPitch;
    private boolean cursorHidden = false;
    private boolean lockCursor = false;

    protected ScreenGUI() {
        super();
        elements = new ArrayList<>();
        Cursor cursor = null;
        try {
            cursor = new Cursor(1, 1, 0, 0, 1, BufferUtils.createIntBuffer(1), null);
        } catch (LWJGLException e) {
            e.printStackTrace();
        } finally {
            emptyCursor = cursor;
        }
    }

    @Override
    public void initGui() {
        if (CURSOR_STATUS != CursorStatus.DEFAULT) hideCursor();

        super.initGui();
        elements.clear();
        init();
    }

    protected void init() {
        if (mc.thePlayer != null) {
            rotationYaw = new float[]{mc.thePlayer.rotationYaw};
            rotationPitch = new float[]{mc.thePlayer.rotationPitch};
        }
    }

    private int getCursorX() {
        if (OptionCore.CURSOR_TOGGLE.getValue()) return lockCursor ? 0 : (width / 2 - mouseX) / 2;
        else return !super.isCtrlKeyDown() ? (width / 2 - mouseX) / 2 : 0;
    }

    private int getCursorY() {
        if (OptionCore.CURSOR_TOGGLE.getValue()) return lockCursor ? 0 : (height / 2 - mouseY) / 2;
        else return !super.isCtrlKeyDown() ? (height / 2 - mouseY) / 2 : 0;
    }

    @Override
    public int getX(boolean relative) {
        return getCursorX();
    }

    @Override
    public int getY(boolean relative) {
        return getCursorY();
    }

    @Override
    public void updateScreen() {
        if (this.elements == null) return;
        for (int i = elements.size() - 1; i >= 0; i--) {
            if (elements.get(i).removed()) {
                elements.get(i).close(mc);
                elements.remove(i);
                continue;
            }

            elements.get(i).update(mc);
        }
    }

    @Override
    public void drawScreen(int cursorX, int cursorY, float f) {
        if (this.elements == null) return;
        for (Elements el : this.elements) if (el == null) return;
        mouseX = cursorX;
        mouseY = cursorY;

        if (mc.thePlayer != null) {
            mc.thePlayer.rotationYaw = rotationYaw[0] - getCursorX() * ROTATION_FACTOR;
            mc.thePlayer.rotationPitch = rotationPitch[0] - getCursorY() * ROTATION_FACTOR;
        }

        super.drawScreen(cursorX, cursorY, f);

        GLCore.glStartUI(mc);

        for (int i = elements.size() - 1; i >= 0; i--) elements.get(i).draw(mc, cursorX, cursorY);

        if (CURSOR_STATUS == CursorStatus.SHOW) {

            GLCore.glBlend(true);
            GLCore.tryBlendFuncSeparate(770, 771, 1, 0);
            GLCore.glBindTexture(OptionCore.SAO_UI.getValue() ? StringNames.gui : StringNames.guiCustom);


            if (mouseDown != 0) {
                final float fval = f * 0.1F;

                if (mouseDownValue + fval < 1.0F) mouseDownValue += fval;
                else mouseDownValue = 1.0F;

                GLCore.glColorRGBA(ColorUtil.CURSOR_COLOR.multiplyAlpha(mouseDownValue));
                GLCore.glTexturedRect(cursorX - 7, cursorY - 7, 35, 115, 15, 15);

                GLCore.glColorRGBA(ColorUtil.DEFAULT_COLOR);
            } else {
                mouseDownValue = 0;

                GLCore.glColorRGBA(ColorUtil.CURSOR_COLOR);
            }

            GLCore.glTexturedRect(cursorX - 7, cursorY - 7, 20, 115, 15, 15);
            GLCore.glBlend(false);
        }
        GLCore.glEndUI(mc);
    }

    @Override
    protected void keyTyped(char ch, int key) throws IOException {
        if (OptionCore.CURSOR_TOGGLE.getValue() && super.isCtrlKeyDown()) lockCursor = !lockCursor;
        super.keyTyped(ch, key);

        elements.stream().filter(element -> element.focus && element.keyTyped(mc, ch, key)).forEach(element -> actionPerformed(element, Actions.KEY_TYPED, key));
    }

    // TODO: check the way elements is built... Breakpoint gives some weird result (at least for base menu)
    @Override
    protected void mouseClicked(int cursorX, int cursorY, int button) throws IOException {
        super.mouseClicked(cursorX, cursorY, button);
        mouseDown |= (0x1 << button);

        boolean clickedElement = false;

        for (int i = elements.size() - 1; i >= 0; i--) {
            if (i >= elements.size()) {
                if (elements.size() > 0) i = elements.size() - 1;
                else break;
            }

            if (elements.get(i).mouseOver(cursorX, cursorY)) {
                if (elements.get(i).mousePressed(mc, cursorX, cursorY, button))
                    actionPerformed(elements.get(i), Actions.getAction(button, true), button);

                clickedElement = true;
            }
        }

        if (!clickedElement) backgroundClicked(cursorX, cursorY, button);
    }

    @Override
    protected void mouseReleased(int cursorX, int cursorY, int button) {
        super.mouseReleased(cursorX, cursorY, button);
        mouseDown &= ~(0x1 << button);

        for (int i = elements.size() - 1; i >= 0; i--) {
            if (i >= elements.size()) {
                if (elements.size() > 0) i = elements.size() - 1;
                else break;
            }
            if (elements.get(i).mouseOver(cursorX, cursorY, button) && elements.get(i).mouseReleased(mc, cursorX, cursorY, button))
                actionPerformed(elements.get(i), Actions.getAction(button, false), button);
        }
    }

    protected void backgroundClicked(int cursorX, int cursorY, int button) {
    }

    private void mouseWheel(int cursorX, int cursorY, int delta) {
        elements.stream().filter(element -> element.mouseOver(cursorX, cursorY) && element.mouseWheel(mc, cursorX, cursorY, delta)).forEach(element -> actionPerformed(element, Actions.MOUSE_WHEEL, delta));
    }

    @Override
    public void actionPerformed(Elements element, Actions action, int data) {
        element.click(mc.getSoundHandler(), false);
    }

    @Override
    public void handleMouseInput() throws IOException {
        super.handleMouseInput();

        if (Mouse.hasWheel()) {
            final int x = Mouse.getEventX() * width / mc.displayWidth;
            final int y = height - Mouse.getEventY() * height / mc.displayHeight - 1;
            final int delta = Mouse.getEventDWheel();

            if (delta != 0) mouseWheel(x, y, delta);
        }
    }

    @Override
    public boolean doesGuiPauseGame() {
        return OptionCore.GUI_PAUSE.getValue();
    }

    @Override
    public void onGuiClosed() {
        showCursor();

        close();
    }

    protected void close() {
        elements.stream().forEach(el -> el.close(mc));
        elements.clear();
    }

    protected void hideCursor() {
        if (!cursorHidden) toggleHideCursor();
    }

    protected void showCursor() {
        if (cursorHidden) toggleHideCursor();
    }

    protected void toggleHideCursor() {
        cursorHidden = !cursorHidden;
        try {
            Mouse.setNativeCursor(cursorHidden ? emptyCursor : null);
        } catch (LWJGLException ignored) {
        }
    }
}

File: src/main/java/com/saomc/saoui/neo/screens/old/ScreenGUI.kt
package com.saomc.saoui.neo.screens.old

import be.bluexin.saomclib.profile
import com.saomc.saoui.GLCore
import com.saomc.saoui.SAOCore
import com.saomc.saoui.api.elements.CategoryEnum
import com.saomc.saoui.api.elements.MenuElementParent
import com.saomc.saoui.api.screens.Actions
import com.saomc.saoui.colorstates.CursorStatus
import com.saomc.saoui.config.OptionCore
import com.saomc.saoui.resources.StringNames
import com.saomc.saoui.themes.elements.menus.CategoryData
import com.saomc.saoui.util.ColorUtil
import net.minecraft.client.gui.GuiScreen
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import org.lwjgl.BufferUtils
import org.lwjgl.input.Cursor
import org.lwjgl.input.Mouse
import java.io.IOException
import java.util.*
import javax.annotation.OverridingMethodsMustInvokeSuper

/**
 * Part of saoui by Bluexin.
 *
 * @author Bluexin
 */
@SideOnly(Side.CLIENT)
abstract class ScreenGUI : GuiScreen(), MenuElementParent {

    companion object {
        protected val ROTATION_FACTOR = 0.25f
        protected var CURSOR_STATUS = CursorStatus.SHOW
        protected val CURSOR_EMPTY = Cursor(1, 1, 0, 0, 1, BufferUtils.createIntBuffer(1), null)
//        protected lateinit var CURSOR_CUSTOM: Cursor
    }

    protected var mouseX = 0
    protected var mouseY = 0
    protected var mouseDown = 0
    protected var mouseDownValue = 0.0F
    protected var rotationYaw = 0.0F
    protected var rotationPitch = 0.0F
    protected var cursorHidden = false
    protected var lockCursor = false
    protected val categories = mutableListOf<CategoryData>()

    @OverridingMethodsMustInvokeSuper
    override fun initGui() {
        if (CURSOR_STATUS != CursorStatus.DEFAULT) hideCursor()
        rotationYaw = mc.player!!.rotationYaw
        rotationPitch = mc.player!!.rotationPitch
        categories.forEach { it.init(this) }

        super.initGui()
    }

    private fun getCursorX(): Int {
        return if (lockCursor) 0 else (width / 2 - mouseX) / 2
    }

    private fun getCursorY(): Int {
        return if (lockCursor) 0 else (height / 2 - mouseY) / 2
    }


    private fun hideCursor() {
        if (!cursorHidden) toggleHideCursor()
    }

    private fun toggleHideCursor() {
        cursorHidden = !cursorHidden
        Mouse.setNativeCursor(if (cursorHidden) CURSOR_EMPTY else null)
    }

    override fun updateScreen() {
    }

    override fun drawScreen(cursorX: Int, cursorY: Int, partialTicks: Float) {
        mc.profile("saoui menu") {
            mouseX = cursorX
            mouseY = cursorY
            mc.player.rotationYaw = rotationYaw - getCursorX() * ROTATION_FACTOR
            mc.player?.rotationPitch = rotationPitch - getCursorY() * ROTATION_FACTOR

            categories.forEach { it.draw(mc, cursorX, cursorY) }

            if (CURSOR_STATUS == CursorStatus.SHOW) { // TODO: maybe there's a way to move all of this to the actual org.lwjgl.input.Cursor

                GLCore.glBlend(true)
                GLCore.tryBlendFuncSeparate(770, 771, 1, 0)
                GLCore.glBindTexture(if (OptionCore.SAO_UI.isEnabled) StringNames.gui else StringNames.guiCustom)

                if (mouseDown != 0) {
                    val fval = partialTicks * 0.1f

                    if (mouseDownValue + fval < 1.0f)
                        mouseDownValue += fval
                    else
                        mouseDownValue = 1.0f

                    GLCore.glColorRGBA(ColorUtil.CURSOR_COLOR.multiplyAlpha(mouseDownValue))
                    GLCore.glTexturedRect((cursorX - 7).toDouble(), (cursorY - 7).toDouble(), 35.0, 115.0, 15.0, 15.0)

                    GLCore.glColorRGBA(ColorUtil.DEFAULT_COLOR)
                } else {
                    mouseDownValue = 0f

                    GLCore.glColorRGBA(ColorUtil.CURSOR_COLOR)
                }

                GLCore.glTexturedRect((cursorX - 7).toDouble(), (cursorY - 7).toDouble(), 20.0, 115.0, 15.0, 15.0)
            }


        }
    }

    override fun closeCategory(category: CategoryEnum) {
        categories.find { it.category == category }?.setEnabled(false)
    }

    override fun openCategory(category: CategoryEnum){
        categories.find { it.category == category }?.setEnabled(true)
    }

    override fun keyTyped(ch: Char, key: Int) {
        if (OptionCore.CURSOR_TOGGLE.isEnabled && GuiScreen.isCtrlKeyDown()) lockCursor = !lockCursor
        SAOCore.LOGGER.debug("ch - $ch key - $key")
        super.keyTyped(ch, key)

//        ElementDispatcher.getElements().stream().anyMatch({ e -> e.keyTyped(mc, ch, key) })

        //elements.menuElements.keySet().stream().filter(Element::isFocus).forEach(element -> actionPerformed(element, Actions.KEY_TYPED, key));
    }

    @Throws(IOException::class)
    override fun mouseClicked(cursorX: Int, cursorY: Int, button: Int) {
        super.mouseClicked(cursorX, cursorY, button)
        mouseDown = mouseDown or (0x1 shl button)
        val mouseButton: Actions = if (button == 0) Actions.LEFT_PRESSED else Actions.RIGHT_PRESSED

        if (categories.firstOrNull { it.mouseClicked(cursorX, cursorY, mouseButton) } == null)
            try {
                backgroundClicked(cursorX, cursorY, mouseButton)
            } catch (e: ConcurrentModificationException) {
                //Do Nothing
                SAOCore.LOGGER.debug("mouseClicked ended unexpectedly")
            }

    }

    override fun mouseReleased(cursorX: Int, cursorY: Int, button: Int) {
        super.mouseReleased(cursorX, cursorY, button)
        val mouseButton: Actions = if (button == 0) Actions.LEFT_RELEASED else Actions.RIGHT_RELEASED
        categories.firstOrNull { it.mouseClicked(cursorX, cursorY, mouseButton) }
        mouseDown = mouseDown and (0x1 shl button).inv()
    }

    open fun backgroundClicked(cursorX: Int, cursorY: Int, button: Actions) {
        SAOCore.LOGGER.debug("Background Clicked")
    }

    private fun mouseWheel(cursorX: Int, cursorY: Int, delta: Int) {
        categories.forEach { it.mouseScroll(cursorX, cursorY, delta) }
    }

    override fun handleMouseInput() {
        super.handleMouseInput()

        if (Mouse.hasWheel()) {
            val x = Mouse.getEventX() * width / mc.displayWidth
            val y = height - Mouse.getEventY() * height / mc.displayHeight - 1
            val delta = Mouse.getEventDWheel()

            if (delta != 0) mouseWheel(x, y, delta)
        }
    }

    override fun doesGuiPauseGame() = false//OptionCore.GUI_PAUSE.isEnabled

    override fun onGuiClosed() {
        showCursor()

        close()
    }

    protected open fun close() = categories.forEach { it.close() }

    private fun showCursor() {
        if (cursorHidden) toggleHideCursor()
    }

    override val parentX: Int
        get() = getCursorX()

    override val parentY: Int
        get() = getCursorY()

    override val parentZ = 0

    operator fun plusAssign(category: CategoryData) {
        this.categories += category
    }

    operator fun CategoryData.unaryPlus() {
        this@ScreenGUI += this
    }
}
